### ISSUE TITLE
Distinct, repeatable(every) job instances of the same task method

### ISSUE THREAD

**Issue by elssar** [rschmukler/agenda#209]
I need agenda to run the same task at separate times each day, with different data being passed to the task at each run. `agenda.every` is almost perfect for that, except that the same task cannot be repeated, as `every` sets the task type as `single`

To get around that limitation, I started dynamically defining tasks, as and when they are added

```
function addTask (params) {
    agenda.define(params.name, params.config, function (job, done) {
        baseTasks[params.baseTask](job.attrs.data, done);
    });
};
```

This worked fine until an exception caused the server to restart and all the task definitions were lost.

So I tried defining all tasks when the server starts, basically by iterating over jobs in the database, and defining tasks for them. This isn't ideal and caused a lot of bugs.

Finally I just started defining tasks myself, simply copying what `agenda.every` does, omitting setting the type to single.

Would be nice to have this by default.

**Comment by loris** [comment]
What kind of data is being passed at each run? Is that information from the previous run?
I do have that use case which is working fine, at the end of a run, I save the information from the run (mainly some cursor IDs where the run finished) to the job data so that the next run will use those data to know where to start from.

**Comment by elssar** [comment]
@loris the data is different for each run, and new jobs can be added dynamically through an API.

**Comment by coderfox** [comment]
:+1: Such features is quite useful!

**Comment by amey1109m** [comment]
I am creating 2 jobs for scheduling auctions to start and close at a particular datetime. Also creating 2  jobs for starting and closing numbers in an auction to start at particular datetime. Currently I am defining dynamic job names. Kindly suggest how to proceed with scheduling with the saved jobs after node server restart with the current approach. The admin can update the auction date time before the auction starts. Code looks like follows

agenda.define('start-'+auctionName,function(job,done){
             myModel.findOneAndUpdate(query,update,function(err,data){
});
});
agenda.schedule(auctionStartDateTime, 'start-'+auctionName);
agenda.start();

agenda.define('end-'+auctionName,function(job,done){
             myModel.findOneAndUpdate(query,update,function(err,data){
});
});
agenda.schedule(auctionStartDateTime, 'end-'+auctionName);
agenda.start();

**Comment by JonnyBGod** [comment]
This features is imperative for my use case similar to @coderfox's.

My idea for this feature to be useful would be to have the chance to define jobs with namespaces like:

```javascript
agenda.define('userJob:*', function(job, done) {
    var userId = job.attrs.data.userId;
});
```

Then just define multiple user jobs like: 

```javascript
agenda.every('30 minutes', 'userJob:user1', { userId: "1"});
agenda.every('20 minutes', 'userJob:user2', { userId: "2"});
```

I will look into your code to see if I can quickly implement this feature if you agree with this behaviour.

**Comment by noisyox** [comment]
@JonnyBGod @loris  What is the status on this feature as I have the same specific requirement of scheduling same task for different users. but right now it overwrites the previous task and does not let me add more similar tasks.

**Comment by BlackHawkNL** [comment]
A feature like this would be awesome, any news on it yet?

**Comment by arihantdaga** [comment]
Any Update on this.. ?
I think it's very important for such good plugin to have functionality for defining the job's behaviour only once but executing it at different times for different data.

**Comment by thardy** [comment]
I've been digging into this issue and found https://github.com/agenda/agenda/issues/410.  The comment by lushc about a third of the way down has been VERY helpful.  I am making progress on this now due to the fact I can finally see separate jobs being persisted with the same name and different data.  

"Every" persists a doc with a type of "single", but the method recommended in that comment persists docs with a type of "normal".  They all share the same job name, but they have different data, which is exactly what I need.  I'm not finished yet, but this looks very promising.

**Comment by HailoMYM** [comment]
any update on this?

**Comment by nagaraja1985** [comment]
Any update on this?

**Comment by djibomar** [comment]
Currently facing a similar issue. Any update?

**Comment by reevr** [comment]
you can use 'create' function in agenda. 
` 

        agenda.define('task-handler', config, function (job, done) {
             // execute job
             done()
        });    
        const job = agenda.create('task-handler', data);
        job.repeatEvery('10 seconds');
        
        await job.save();`

**Comment by irfhanhilaly** [comment]
adding solution found by @thardy in #410 here for quick reference.

// basically agenda.every() but without setting the job type to "single" & multiple definition scheduling
const every = (interval, name, data, options) => {
  return new Promise((resolve, reject) => {
    agenda.create(name, data)
      //.schedule(interval) if you don't want the job to run right away
      .repeatEvery(interval, options)
      .save((err) => err ? reject(err) : resolve());
  });
};
